### PR TITLE
Load plugins deterministically. Fixes #1010.

### LINKED ISSUE #1010: Clash renames are non-deterministic
In order to let us help you better, please fill out the following fields as best you can:

### I am...

* [x] Reporting a bug
* [ ] Suggesting a new feature
* [x] Requesting help with running my bot
* [ ] Requesting help writing plugins
* [ ] Here about something else

### I am running...

* Errbot version: 5.0.1
* OS version: Debian Jessie
* Python version: 3.4.2
* Using a virtual environment: Docker (https://hub.docker.com/r/rroemhild/errbot/)

### Issue description

I have a plugin with a `status` command defined. This of course clashes with the core `status` command. It is unknown until start-up if my `status` gets renamed to `myplugin-status` or the core to `health-status`.

### Steps to reproduce

Just create a 'helloworld' plugin but call the function `status`...

    $ errbot -T
    ...
    HelloWorld.status clashes with Health.status so it has been renamed hellowworld-status
    ...

Run it a few times and eventually it will respond differently

    $ errbot -T
    ...
    Health.status clashes with HelloWorld.status so it has been renamed health-status
    ...
    

### Additional info

Is there a way for my plugin to ensure it gets the preferred command name?
For now, I just use a different command name :)

### PR DESCRIPTION
**Pull request by KimNorgaard**

### CODE REVIEW

**Comment by KimNorgaard** [comment]
Test error is caused by #1249.

**Comment by GenPage** [comment]
Hopefully @gbin will get #1249 addressed soon

File: errbot/utils.py [hunk 1]
```diff
@@ -146,7 +147,7 @@ def find_roots(path, file_sig='*.plug'):
             the file pattern to look for
        :return: a set of paths
     """
-    roots = set()  # you can have several .plug per directory.
```
Comment (gbin-argo) [inline errbot/utils.py]: No, you have ignored the comment, you need to setup them right?

Comment (KimNorgaard) [reply errbot/utils.py]: I am not sure I understand your question, but let me try to get it right.

Before, a `set` was used to return a unique list of paths in which to look for plugins. The problem with using a `set` is that it is unordered. I replaced the set with a `list`. This means that the paths will be added multiple times (as stated by the comment).

In order to make the list unique again and at the same same preserve order, I use an `OrderedDict`. I moved the original comment to the `return` statement to indicate this:

```python
    return list(collections.OrderedDict.fromkeys(roots))  # you can have several .plug per directory
```

If the purpose of the original `set` was to ensure the list was unique, then this PR should not change that. It just assures the order is always the same.

Am I missing something?

Comment (gbin-argo) [reply errbot/utils.py]: Doh sorry about, thanks for the patch!

File: errbot/utils.py [hunk 2]
```diff
@@ -160,8 +161,8 @@ def find_roots(path, file_sig='*.plug'):
                     log.debug('Ignore %s.', dir_to_add)
                     break
             else:
-                roots.add(dir_to_add)
-    return roots
+                roots.append(dir_to_add)
+    return list(collections.OrderedDict.fromkeys(roots))  # you can have several .plug per directory.s
```
Comment (sijis) [inline errbot/utils.py]: nitpick. There's a typo in the comment `directory.s`.

File: errbot/utils.py [hunk 3]
```diff
@@ -175,13 +176,13 @@ def collect_roots(base_paths, file_sig='*.plug'):
             the file pattern to look for
        :return: a set of paths
```
Comment (sijis) [inline errbot/utils.py]: We should update the method signature.

File: errbot/utils.py [hunk 4]
```diff
@@ -146,7 +147,7 @@ def find_roots(path, file_sig='*.plug'):
...
```
Comment (sijis) [inline errbot/utils.py]: We should update the method docstring.

**Comment by sijis** [comment]
Do you think we should update this line too, to send a list?
https://github.com/errbotio/errbot/blob/8c3d74995e316a14870527901ce89f2733691c53/errbot/backend_plugin_manager.py#L28

Its the only reference i see for `collect_roots()` used outside of `utils.py`.

**Comment by sijis** [comment]
Otherwise, I think this PR looks great! Awesome! :+1:

**Comment by GenPage** [comment]
@gbin-argo Need your approval now as well before merging

**Comment by sijis** [comment]
I'm closing this in favor of https://github.com/errbotio/errbot/pull/1282. Its based off this work but it has additional feedback from other contributors.

Thank you for initiating these fixes.

**Pull request closed without merging**